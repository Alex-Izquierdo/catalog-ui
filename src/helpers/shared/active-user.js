/* eslint camelcase: 0 */
import { getAxiosInstance } from '../shared/user-login';
import { CATALOG_API_BASE } from '../../utilities/constants';
const axiosInstance = getAxiosInstance();

export const getUser = () => axiosInstance.get(`${CATALOG_API_BASE}/me/`);
