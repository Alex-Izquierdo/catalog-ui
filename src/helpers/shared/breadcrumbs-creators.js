export const BASE_PORTFOLIO_FRAGMENTS = [
  {
    title: 'Portfolios',
    pathname: '/portfolios',
    searchParams: {}
  }
];

export const ENTITIES_EXTRA_PARAMS = {
  'portfolio-item': ['source']
};

export const FRAGMENT_TITLE = {
  '/portfolio': (getState) =>
    getState().portfolioReducer.selectedPortfolio.name,
  '/portfolio/portfolio-item': (getState) =>
    getState().portfolioReducer.portfolioItem.portfolioItem.name,
  '/portfolio/portfolio-item/edit-survey': () => 'Edit survey',
  '/portfolio/portfolio-item/add-products': () => 'Add products',
  '/platform': (getState) => getState().platformReducer.selectedPlatform.name,
  '/platform/platform-templates': () => 'Templates',
  '/platform/platform-inventories': () => 'Inventories',
  '/platform/service-offerings': (getState) =>
    getState().platformReducer.serviceOffering.service.name,
  '/order': (getState) => {
    const { portfolioItem, order } = getState().orderReducer.orderDetail;
    return `${portfolioItem.name} # ${order.id}`;
  }
};

export const FRAGMENT_PREFIX = {
  '/portfolio': {
    pathname: '/portfolios',
    title: 'Portfolios',
    searchParams: {}
  },
  '/platform': {
    pathname: '/platforms',
    title: 'Platforms',
    searchParams: {}
  },
  '/order': {
    pathname: '/orders',
    title: 'Orders',
    searchParams: {}
  }
};
