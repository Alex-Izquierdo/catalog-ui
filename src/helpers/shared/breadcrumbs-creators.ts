import {
  PORTFOLIOS_ROUTE,
  PORTFOLIO_ROUTE,
  PORTFOLIO_ITEM_ROUTE,
  ORDER_ROUTE,
  PLATFORM_ROUTE,
  PLATFORM_SERVICE_OFFERINGS_ROUTE,
  PLATFORMS_ROUTE,
  ORDERS_ROUTE,
  EDIT_SURVEY_ROUTE,
  ADD_PRODUCTS_ROUTE,
  PORTFOLIO_ITEM_ROUTE_EDIT
} from '../../constants/routes';
import { PortfolioReducerState } from '../../redux/reducers/portfolio-reducer';
import { ReactNode } from 'react';
import { PlatformReducerState } from '../../redux/reducers/platform-reducer';
import { OrderReducerState } from '../../redux/reducers/order-reducer';

export const BASE_PORTFOLIO_FRAGMENTS = [
  {
    title: 'Portfolios',
    pathname: PORTFOLIOS_ROUTE,
    searchParams: {}
  }
];

export const ENTITIES_EXTRA_PARAMS = {
  'portfolio-item': ['source']
};

export const FRAGMENT_TITLE = {
  ['/portfolio']: (
    getState: () => { portfolioReducer: PortfolioReducerState }
  ): ReactNode => getState().portfolioReducer.selectedPortfolio.name,
  ['/portfolio/portfolio-item']: (
    getState: () => { portfolioReducer: PortfolioReducerState }
  ): ReactNode => getState().portfolioReducer.portfolioItem.portfolioItem.name,
  [EDIT_SURVEY_ROUTE]: (): string => 'Edit survey',
  [ADD_PRODUCTS_ROUTE]: (): string => 'Add products',
  [PLATFORM_ROUTE]: (
    getState: () => { platformReducer: PlatformReducerState }
  ): ReactNode => getState().platformReducer.selectedPlatform.name,
  [PLATFORM_SERVICE_OFFERINGS_ROUTE]: (
    getState: () => { platformReducer: PlatformReducerState }
  ): ReactNode => getState().platformReducer.serviceOffering.service.name,
  [ORDER_ROUTE]: (
    getState: () => { orderReducer: OrderReducerState }
  ): ReactNode => {
    const { portfolioItem, order } = getState().orderReducer.orderDetail;
    return `${portfolioItem.name} # ${order.id}`;
  },
  [PORTFOLIO_ITEM_ROUTE_EDIT]: (): string => 'Edit product'
};

export const FRAGMENT_PREFIX = {
  [PORTFOLIO_ROUTE]: {
    pathname: PORTFOLIOS_ROUTE,
    title: 'Portfolios',
    searchParams: {}
  },
  [PLATFORM_ROUTE]: {
    pathname: PLATFORMS_ROUTE,
    title: 'Platforms',
    searchParams: {}
  },
  [ORDER_ROUTE]: {
    pathname: ORDERS_ROUTE,
    title: 'Orders',
    searchParams: {}
  }
};
