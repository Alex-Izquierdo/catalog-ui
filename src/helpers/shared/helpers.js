import moment from 'moment';

export const scrollToTop = () => document.getElementById('root').scrollTo({
  behavior: 'smooth',
  top: 0,
  left: 0
});

export const filterServiceOffering = ({ display_name, name }, filter) => {
  const filterAtrribute = display_name || name;
  return filterAtrribute.trim().toLowerCase().includes(filter.toLowerCase());
};

export const allowNull = wrappedPropTypes => (props, propName, ...rest) => {
  if (props[propName] === null) {
    return null;
  }

  return wrappedPropTypes(props, propName, ...rest);
};

const oneDay = 24 * 60 * 60 * 1000;
export const calcuateDiffDays = (firstDate, secondDate) => Math.round(Math.abs((firstDate.getTime() - secondDate.getTime()) / (oneDay)));

export const createModifiedLabel = (date, user) => `Last modified ${calcuateDiffDays(new Date(), date)} days ago${ user ? ` by ${user}.` : '.'  }`;

export const createOrderedLabel = date => {
  const orderedAgo = calcuateDiffDays(date, new Date());
  return `Ordered ${orderedAgo} ${orderedAgo > 1 ? 'days' : 'day'} ago`;
};

export const createUpdatedLabel = orderItem => {
  if (!orderItem[0]) {
    return null;
  };

  const orderedAgo = calcuateDiffDays(new Date(orderItem[0].updated_at), new Date());
  return `Updated ${orderedAgo} ${orderedAgo > 1 ? 'days' : 'day'} ago`;
};

export const createDateString = date => moment(new Date(date).toUTCString(), 'DD-MMM-YYYY, HH:mm').format('DD MMM YYYY, HH:mm UTC');
