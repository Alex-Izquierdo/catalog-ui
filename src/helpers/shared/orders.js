import { CATALOG_API_BASE } from '../../utilities/constants';

export const getOrderIcon = ({ orderItems }) =>
  orderItems[0] &&
  `${CATALOG_API_BASE}/portfolio_items/${orderItems[0].portfolio_item_id}/icon`;

export const getOrderPortfolioName = ({ orderItems, id }, portfolioItems) => {
  const portfolioItem =
    orderItems[0] &&
    portfolioItems.find(({ id }) => orderItems[0].portfolio_item_id === id);
  return portfolioItem ? portfolioItem.name : `Order ${id}`;
};

export const getOrderPlatformId = ({ orderItems }, portfolioItems) => {
  const portfolioItem =
    orderItems[0] &&
    portfolioItems.find(({ id }) => orderItems[0].portfolio_item_id === id);
  return portfolioItem
    ? {
        orderPlatform: portfolioItem.service_offering_source_ref,
        orderPortfolio: portfolioItem.portfolio_id
      }
    : {};
};
