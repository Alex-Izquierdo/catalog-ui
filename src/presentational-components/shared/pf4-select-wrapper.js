import React from 'react';
import PropTypes from 'prop-types';
import { FormSelect, FormSelectOption, FormGroup, TextContent, Text, TextVariants } from '@patternfly/react-core';

const createOptions = (options, inputValue, isRequired) => {
  if (inputValue && isRequired) {
    return options;
  }

  let selectOptions = [ ...options ];
  return selectOptions.find(({ value }) => value === undefined)
    ? [ ...selectOptions ]
    : [{ label: isRequired ? 'Please choose' : 'None' }, ...selectOptions ];
};

const Select = ({
  input,
  options,
  isReadOnly,
  isDisabled,
  FieldProvider,
  isRequired,
  formOptions: { change },
  ...rest
}) => (
  <FormSelect { ...input } { ...rest }
    onChange={ (value, ...args) => {
      if (rest.onChange) {
        rest.onChange(value);
        change(input.name, value);
      } else {
        input.onChange(value, ...args);
      }
    } } isDisabled={ isDisabled || isReadOnly }>
    { createOptions(options, input.value, isRequired).map((props) => (
      <FormSelectOption
        key={ props.value || props.label } // eslint-disable-line react/prop-types
        { ...props }
        label={ props.label.toString() }
      />
    )) }
  </FormSelect>
);

Select.propTypes = {
  input: PropTypes.object.isRequired,
  options: PropTypes.arrayOf(PropTypes.shape({
    value: PropTypes.any,
    label: PropTypes.string.isRequired
  })).isRequired,
  isReadOnly: PropTypes.bool,
  isDisabled: PropTypes.bool,
  isRequired: PropTypes.bool,
  FieldProvider: PropTypes.any,
  formOptions: PropTypes.shape({
    change: PropTypes.func
  })
};

Select.defaultProps = {
  formOptions: {}
};

const Pf4SelectWrapper = ({
  componentType,
  label,
  isRequired,
  helperText,
  meta,
  description,
  hideLabel,
  formOptions,
  dataType,
  initialKey,
  id,
  ...rest
}) => {
  const { error, touched } = meta;
  const showError = touched && error;
  const { name } = rest.input;

  return (
    <FormGroup
      isRequired={ isRequired }
      label={ !hideLabel && label }
      fieldId={ id || name }
      isValid={ !showError }
      helperText={ helperText }
      helperTextInvalid={ meta.error }
    >
      { description && <TextContent><Text component={ TextVariants.small }>{ description }</Text></TextContent> }
      <Select formOptions={ formOptions } id={ id || name } label={ label } isValid={ !showError } isRequired={ isRequired } { ...rest }/>
    </FormGroup>
  );
};

Pf4SelectWrapper.propTypes = {
  componentType: PropTypes.string,
  id: PropTypes.string,
  label: PropTypes.string,
  isRequired: PropTypes.bool,
  helperText: PropTypes.string,
  meta: PropTypes.object,
  description: PropTypes.string,
  hideLabel: PropTypes.bool,
  formOptions: PropTypes.object,
  dataType: PropTypes.string,
  initialKey: PropTypes.any
};

export default Pf4SelectWrapper;
