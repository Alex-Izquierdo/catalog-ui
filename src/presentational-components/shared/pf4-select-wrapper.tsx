/* eslint-disable react/prop-types */
import React, { useState, useRef, ReactNode } from 'react';
import useFieldApi, {
  ValidatorType
} from '@data-driven-forms/react-form-renderer/dist/cjs/use-field-api';
import {
  InternalSelect,
  SelectOption
} from '@data-driven-forms/pf4-component-mapper/dist/cjs/select';
import {
  FormGroup,
  TextContent,
  Text,
  TextVariants
} from '@patternfly/react-core';
import useFormApi from '@data-driven-forms/react-form-renderer/dist/cjs/use-form-api';
import { defineMessage } from 'react-intl';
import useFormatMessage from '../../utilities/use-format-message';
import { SelectOptions } from '../../types/common-types';

const createOptions = (
  options: SelectOptions,
  inputValue: string,
  isRequired: boolean,
  optionsMessages: { choose: ReactNode; none: ReactNode }
): SelectOption[] => {
  if (inputValue && isRequired) {
    return options as SelectOption[];
  }

  const selectOptions = [...options];
  return selectOptions.find(({ value }) => value === undefined)
    ? ([...selectOptions] as SelectOption[])
    : ([
        { label: isRequired ? optionsMessages.choose : optionsMessages.none },
        ...selectOptions
      ] as SelectOption[]);
};

interface SelectProps {
  input: {
    name: string;
    onChange: (value: any) => void;
    value: any;
  };
  id?: string;
  label?: ReactNode;
  options?: SelectOptions;
  isReadOnly?: boolean;
  isDisabled?: boolean;
  isRequired?: boolean;
  validated?: 'default' | 'success' | 'error';
  multi?: boolean;
  loadOptions?: (
    search?: string,
    lookupArguments?: any[]
  ) => Promise<SelectOption[]>;
  meta: { initial?: any };
  onChange?: (...args: any[]) => unknown;
}
const Select: React.ComponentType<SelectProps> = ({
  input,
  options = [],
  isDisabled = false,
  isRequired = false,
  multi = false,
  loadOptions,
  meta,
  ...rest
}) => {
  const formatMessage = useFormatMessage();
  const optionsMessages = useRef<{
    none: ReactNode;
    choose: ReactNode;
  }>({
    none: formatMessage(
      defineMessage({
        id: 'forms.select.options.none',
        defaultMessage: 'None'
      })
    ),
    choose: formatMessage(
      defineMessage({
        id: 'forms.select.options.choose',
        defaultMessage: 'Please choose'
      })
    )
  });
  const [initialFetch, setInitialFetch] = useState(true);
  const formOptions = useFormApi();
  let loadOptionsOverride = loadOptions;
  if (loadOptions && meta.initial) {
    const lookupArguments = Array.isArray(meta.initial)
      ? meta.initial.map((option) =>
          typeof option === 'object' ? option.value : option
        )
      : [meta.initial];
    loadOptionsOverride = (filterValue) => {
      return initialFetch
        ? loadOptions(filterValue, lookupArguments).then((initialOptions) => {
            return loadOptions(filterValue).then((options) => {
              setInitialFetch(false);
              return [
                ...initialOptions,
                ...options.filter(
                  ({ value }) =>
                    !initialOptions.find((option) => option.value === value)
                )
              ];
            });
          })
        : loadOptions(filterValue);
    };
  }

  return (
    <InternalSelect
      menuIsPortal
      {...input}
      {...rest}
      loadOptions={loadOptionsOverride}
      onChange={(value, ...args) => {
        if (rest.onChange) {
          rest.onChange(value);
          formOptions.change(input.name, value);
        } else {
          input.onChange(value, ...args);
        }
      }}
      isMulti={multi}
      options={
        !loadOptions
          ? createOptions(
              options,
              input.value,
              isRequired,
              optionsMessages.current
            )
          : undefined
      }
      isDisabled={isDisabled}
    />
  );
};

export interface Pf4SelectWrapperProps
  extends Omit<Omit<SelectProps, 'input'>, 'meta'> {
  name: string;
  label?: ReactNode;
  isRequired?: boolean;
  helperText?: ReactNode;
  description?: ReactNode;
  hideLabel?: boolean;
  id?: string;
  initialValue?: any;
  validate?: ValidatorType[];
}
const Pf4SelectWrapper: React.ComponentType<Pf4SelectWrapperProps> = (
  props
) => {
  const {
    label,
    isRequired,
    helperText,
    meta,
    description,
    hideLabel,
    id,
    ...rest
  } = useFieldApi(props);
  const { error, touched } = meta;
  const showError = touched && error;
  const { name } = rest.input;

  return (
    <FormGroup
      isRequired={isRequired}
      label={!hideLabel && label}
      fieldId={id || name}
      validated={showError ? 'error' : 'default'}
      helperText={helperText}
      helperTextInvalid={meta.error}
    >
      {description && (
        <TextContent>
          <Text component={TextVariants.small}>{description}</Text>
        </TextContent>
      )}
      <Select
        id={id || name}
        meta={meta}
        label={label}
        validated={showError ? 'error' : 'default'}
        isRequired={isRequired}
        {...rest}
      />
    </FormGroup>
  );
};

export default Pf4SelectWrapper;
