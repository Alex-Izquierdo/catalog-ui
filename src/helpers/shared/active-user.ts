/* eslint camelcase: 0 */
import { getAxiosInstance } from '../shared/user-login';
import {
  CATALOG_API_BASE,
  EXTERNAL_LOGIN_URI
} from '../../utilities/constants';
const axiosInstance = getAxiosInstance();

export const getUser = () => axiosInstance.get(`${CATALOG_API_BASE}/me/`);

export const logoutUser = () => {
  localStorage.removeItem('catalog_user');
  return axiosInstance.post(`${CATALOG_API_BASE}/logout/`);
};

export const loginUser = () => {
  window.location.replace(EXTERNAL_LOGIN_URI);
  return null;
};
