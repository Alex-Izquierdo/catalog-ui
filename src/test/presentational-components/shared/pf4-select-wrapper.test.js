import React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import { Pf4SelectWrapper } from '../../../presentational-components/shared/pf4-select-wrapper';
import { InternalSelect } from '@data-driven-forms/pf4-component-mapper/select';
import Form from '@data-driven-forms/react-form-renderer/form';

describe('<Pf4SelectWrapper />', () => {
  let initialProps;
  beforeEach(() => {
    initialProps = {
      id: 'bazz',
      name: 'bazz',
      options: [
        {
          label: 'Foo',
          value: 'bar'
        }
      ]
    };
  });

  it('should render correctly', () => {
    const wrapper = shallow(
      <Form onSubmit={Function}>
        {() => <Pf4SelectWrapper {...initialProps} />}
      </Form>
    ).dive();
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should create empty option', () => {
    const wrapper = mount(
      <Form onSubmit={Function}>
        {() => <Pf4SelectWrapper {...initialProps} />}
      </Form>
    );
    const options = wrapper.find(InternalSelect).props().options;
    expect(options).toHaveLength(2);
  });

  it('should create empty option for required field', () => {
    const wrapper = mount(
      <Form onSubmit={Function}>
        {() => <Pf4SelectWrapper isRequired {...initialProps} />}
      </Form>
    );
    const options = wrapper.find(InternalSelect).props().options;
    expect(options).toHaveLength(2);
    expect(options[0].label).toEqual('Please choose');
  });

  it('should not create empty option', () => {
    const wrapper = mount(
      <Form onSubmit={Function}>
        {() => (
          <Pf4SelectWrapper
            {...initialProps}
            options={[{ label: 'Foo', value: 'bar' }, { label: 'Empty value' }]}
          />
        )}
      </Form>
    );
    const options = wrapper.find(InternalSelect).props().options;
    expect(options).toHaveLength(2);
  });

  it('should not create empty option if select has value and is required', () => {
    const wrapper = mount(
      <Form onSubmit={Function}>
        {() => (
          <Pf4SelectWrapper
            isRequired
            {...initialProps}
            initialValue="bar"
            options={[{ label: 'Foo', value: 'bar' }]}
          />
        )}
      </Form>
    );
    const options = wrapper.find(InternalSelect).props().options;
    expect(options).toHaveLength(1);
  });

  it('should render correctly in error state', () => {
    const wrapper = shallow(
      <Form onSubmit={Function}>
        {() => (
          <Pf4SelectWrapper
            {...initialProps}
            isRequired
            validate={() => false}
          />
        )}
      </Form>
    ).dive();
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should render correctly with description', () => {
    const wrapper = shallow(
      <Form onSubmit={Function}>
        {() => <Pf4SelectWrapper {...initialProps} description="description" />}
      </Form>
    ).dive();
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
