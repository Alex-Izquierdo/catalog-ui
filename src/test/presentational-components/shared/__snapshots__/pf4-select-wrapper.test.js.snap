// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Pf4SelectWrapper /> should render correctly 1`] = `
<Component
  id="bazz"
  name="bazz"
  options={
    Array [
      Object {
        "label": "Foo",
        "value": "bar",
      },
    ]
  }
/>
`;

exports[`<Pf4SelectWrapper /> should render correctly in error state 1`] = `
<Component
  id="bazz"
  isRequired={true}
  name="bazz"
  options={
    Array [
      Object {
        "label": "Foo",
        "value": "bar",
      },
    ]
  }
  validate={[Function]}
/>
`;

exports[`<Pf4SelectWrapper /> should render correctly with description 1`] = `
<Component
  description="description"
  id="bazz"
  name="bazz"
  options={
    Array [
      Object {
        "label": "Foo",
        "value": "bar",
      },
    ]
  }
/>
`;
